New feature: Allow download from private GitHub repository

minamijoyo opened the pull request:
## Why?
I would like to download the release asset of the private GitHub repository.

I'm using a private brew tap at my company and the prebuild binaries for internal tools are uploaded to the private GitHub Release page.
For such in-house use, S3DownloadStrategy already exists, but most of the source code are managed by GitHub, I think it would be useful to download the binary directly from GitHub.

## What I did
I implemented a new DownloadStrategy to find and download asset using GitHub API.

## How I did it
First, resolve the asset_id associated with the release tag with the following API:
https://developer.github.com/v3/repos/releases/#get-a-release-by-tag-name

and then, download the asset with the following API:
https://developer.github.com/v3/repos/releases/#get-a-single-release-asset

## How to verify it
To use it, add `:using => GitHubReleaseDownloadStrategy` to the URL section of your formula.

```rb
require "formula"

class Hoge < Formula
  url "https://github.com/:owner/:repo/releases/download/:tag/hoge_v0.1.0_darwin_amd64.tar.gz", :using => GitHubReleaseDownloadStrategy
  ...

end
```

```bash
$ export HOMEBREW_GITHUB_API_TOKEN=xxxxxx
$ brew install hoge
```

Please review this.

## Check List
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -532,6 +532,82 @@ def _fetch
+# because it lets you use a private GttHub repository for internal distribution.
+# It works with public one, but in that case simply use CurlDownloadStrategy.
+class GitHubReleaseDownloadStrategy < CurlDownloadStrategy
+  require 'open-uri'
```
This should use `utils/curl` and/or `utils/github` instead.

minamijoyo replied:
Oh, I did not know such a usuful lib. I changed to use `utils/github`.
I also changed the environment varaible used to this same `HOMEBREW_GITHUB_API_TOKEN` .

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -532,6 +532,82 @@ def _fetch
+
+    @github_token = ENV["GITHUB_TOKEN"]
+    unless @github_token
+      puts "Environmental variable GITHUB_TOKEN is required."
```
Perhaps this string should be part of the `raise` instead?

minamijoyo replied:
Sure, I changed it.

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -532,6 +532,82 @@ def _fetch
+
+    url_pattern = %r|https://github.com/(\S+)/(\S+)/releases/download/(\S+)/(\S+)|
+    unless @url =~ url_pattern
+      puts "Invalid url pattern for GitHub Release."
```
Perhaps this string should be part of the raise instead?

minamijoyo replied:
I updated it here as well.

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -532,6 +532,82 @@ def _fetch
+      raise CurlDownloadStrategyError, @url
+    end
+
+    url_pattern = %r|https://github.com/(\S+)/(\S+)/releases/download/(\S+)/(\S+)|
```
This should allow https://github.com/Homebrew/brew/archive/1.1.5.tar.gz format links.

minamijoyo replied:
@MikeMcQuaid I am trying to support this, but the following URL does not contain any asset.
https://api.github.com/repos/Homebrew/brew/releases/tags/1.1.5

How do we download it? I could not find an GitHub API to get archive. Are there any good ideas?

minamijoyo replied:
I've supported both cases.

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -532,6 +532,63 @@ def _fetch
+    return assets.first["id"]
+  end
+
+  def release_url
```
I reckon you can just make this a variable or inline it rather than a function.

minamijoyo replied:
Fixed.

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -532,6 +532,63 @@ def _fetch
+  end
+
+  def _fetch
+    puts "Download asset_id: #{asset_id}"
```
I think this shouldn't be outputted as it's not necessary for the user to see.

minamijoyo replied:
Fixed.

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -532,6 +532,63 @@ def _fetch
+    super
+
+    @github_token = ENV["HOMEBREW_GITHUB_API_TOKEN"]
+    raise CurlDownloadStrategyError, "Environmental variable HOMEBREW_GITHUB_API_TOKEN is required." unless @github_token
```
Use `unless ... end` so this line is shorter.

MikeMcQuaid replied:
Also, you probably need to handle the case where the token doesn't have e.g. private repository access.

minamijoyo replied:
Fixed.

MikeMcQuaid commented:
> @MikeMcQuaid I am trying to support this format, but the following URL does not contain any asset.
> https://api.github.com/repos/Homebrew/brew/releases/tags/1.1.5
> 
> How do we download it? I could not find an GitHub API to get archive. Are there any good ideas?

@minamijoyo I'm thinking this download strategy could be a lot more general and instead be e.g. `GitHubTokenDownloadStrategy`. You can access private resources like if e.g. https://github.com/Homebrew/brew/archive/1.1.5.tar.gz was private using the same token format you've used but without the API call i.e. just a `curl https://$TOKEN@github.com/Homebrew/brew/archive/1.1.5.tar.gz`. That means we can avoid any API calls or parsing and just allow this download strategy to take a token and pass that to GitHub for downloading any private resource. Thoughts?

minamijoyo commented:
@MikeMcQuaid Thanks for the review and advice.
First let me share my current understanding.

Certainly the archive url can be accessed with the curl.
`https://$TOKEN@github.com/:owner/:repo/archive/:filename`

But the releases url is not. In the case of the a private repository, even if you set a token, not found is returned. I don't know why there is such a difference. Try it with a private repository.
`https://$TOKEN@github.com/:owner/:repo/releases/download/:tag/:filename`
 
Although the url of assets API can be accessed with the curl, asset_id is not shown on Web,  I don't know how to get asset_id without API.
`https://$TOKEN@api.github.com/repos/:owner/:repo/releases/assets/:asset_id`

In addition, to download the asset's binary content, we must set the Accept header of the request to `application/octet-stream`.

Therefore it is difficult to make it a common implementation without API calls. Is there a way to go something good? If we assume that both logic are implemented respectively, I think that it is better to separate them into different classes, but one is not what I originally wanted to do.

MikeMcQuaid commented:
@minamijoyo I think it's fine to have API calls if they are needed for releases 👍 but I think this PR should probably handle both cases in order to be useful. The GitHub Releases download strategy could be a subclass from a GitHub Private Repository download strategy to share e.g. at least the token logic.



> But the releases url is not. In the case of the a private repository, even if you set a token, not found is returned. I don't know why there is such a difference.

It may be worth asking support@github.com why this is the case and their thoughts; they will definitely be able to help.

minamijoyo commented:
@MikeMcQuaid OK. I will try to support both cases. Which of the base class names should be `GitHubTokenDownloadStrategy` or `GitHubPrivateRepositoryDownloadStrategy` ?

Also I will also ask GitHub why there is such a difference.

MikeMcQuaid commented:
@minamijoyo Thanks! I think `GitHubPrivateRepositoryDownloadStrategy` and `GitHubPrivateRepositoryReleaseDownloadStrategy` are the best names.

minamijoyo commented:
@MikeMcQuaid OK.  I will do so.

minamijoyo commented:
@MikeMcQuaid I've fixed it, so please review again.
In addition, as the class name changed, I've also changed the title of this pull request to a more appropriate name.

minamijoyo commented:
I contacted GitHub's support and got answers. The summary is as follows:

>> * Is it possible to download private release assets directly with curl without using API?

> No, not currently.

>> * If possible please tell me how to do it.

> It's not possible.

>> * If not, compared to the archive curl, why is there such a difference?

> The fact that it's possible to do for the archive doesn't mean it will be possible for everything else -- you should not have such expectations and I don't believe we'll support that for releases anytime soon. To clarify a bit why those two are different -- those two things are stored differently and go through different paths. In general, you should be using the API for all automation purposes.

After all, we can not get the private release assets without API.

MikeMcQuaid commented:
Fantastic work @minamijoyo. Thanks for your first contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

@joscha @stefansundin This should help you with your use-cases.

joscha commented:
Nice work @minamijoyo!

minamijoyo commented:
@MikeMcQuaid Thank you for review :beer:
